src/index/LocatedTriples.h

```diff
@@ -0,0 +1,174 @@
+
+  // Add `locatedTriples` to the `LocatedTriplesPerBlock`.
+  // Return handles to where they were added (`LocatedTriples` is a sorted set,
+  // see above). We need the handles so that we can easily remove the
+  // `locatedTriples` from the set again in case we need to.
+  //
+  // PRECONDITIONS:
+  //
+  // 1. The `locatedTriples` must not already exist in
+  // `LocatedTriplesPerBlock`.
+  std::vector<LocatedTriples::iterator> add(
+      const std::vector<LocatedTriple>& locatedTriples);
+
+  void erase(size_t blockIndex, LocatedTriples::iterator iter);
+
+  // Get the total number of `LocatedTriple`s (for all blocks).
+  size_t numTriples() const { return numTriples_; }
+
+  // Get the number of blocks with a non-empty set of located triples.
+  size_t numBlocks() const { return map_.size(); }
+
+  // Remove all located triples.
+  void clear() {
+    map_.clear();
+    numTriples_ = 0;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os,
+                                  const LocatedTriplesPerBlock& ltpb) {
+    // Get the block indices in sorted order.
+    std::vector<size_t> blockIndices;
+    std::ranges::transform(ltpb.map_, std::back_inserter(blockIndices),
+                           [](const auto& entry) { return entry.first; });
+    std::ranges::sort(blockIndices);
+    for (auto blockIndex : blockIndices) {
+      os << "LTs in Block #" << blockIndex << ": " << ltpb.map_.at(blockIndex)
+         << std::endl;
+    }
+    return os;
+  }
+};
+
+// Human-readable representation , which are very useful for debugging.
+// TODO<qup42>: find a better place for these definitions
+std::ostream& operator<<(std::ostream& os,
+                         const columnBasedIdTable::Row<Id>& idTableRow);
+std::ostream& operator<<(std::ostream& os, const IdTable& idTable);
+template <typename T, std::size_t N>
+std::ostream& operator<<(std::ostream& os, const std::array<T, N>& v);
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
+
+// DEFINITION OF THE POSITION OF A LOCATED TRIPLE IN A PERMUTATION
+//
+// 1. The position is defined by the index of a block in the permutation and the
+// index of a row within that block.
+//
+// 2. If the triple is contained in the permutation, it is contained exactly
+// once and so there is a well defined block and position in that block.
+//
+// 2. If there is a block, where the first triple is smaller and the last triple
+// is larger, then that is the block and the position in that block is that of
+// the first triple that is (not smaller and hence) larger.
+//
+// 3. If the triple falls "between two blocks" (the last triple of the previous
+// block is smaller and the first triple of the next block is larger), then the
+// position is the first position in that next block.
+//
+// 4. As a special case of 3, if the triple is smaller than all triples in the
+// permutation, the position is the first position of the first block.
```

Comment:
I agree that 5 is a special case that should be highlighted. But imo both are special cases that don't follow from the other rules. If a triple is before/after all triples there are no blocks it falls between, because it is only "neighboring" one block. So both would be undefined as of this definition.

Reply:
So 
2-4 are a complicated way of saying "A triple belongs to the first block (with the smalles index) that also contains at least one triple that is larger than the input triple" (not happy with the formulation).

What is here 2-4 is than an "in particular" explanation of scenarios that follow.

5. Is then the special case that no such block exists, then there is this shadow block.